### PR TITLE
test(ivy): create todo app in ivy

### PR DESCRIPTION
**Pull request by mhevery**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: packages/core/test/bundling/todo/index.html [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      Script tag which bootstraps the application. Use `?debug` in URL to select
+      the debug version of the script.
+      
+      There are two scripts sources: `bundle.min.js` and `bundle.min_debug.js` You can 
+      switch between which bundle the browser loads to experiment with the application.
+
+      - `bundle.min.js`: Is what the site would serve to their users. It has gone
+        through rollup, build-optimizer, and uglify with tree shaking.
+      - `bundle.min_debug.js`: Is what the developer would like to see when debugging
+        the application. It has also done through full pipeline of rollup, build-optimizer,
```
Comment (kara) [inline packages/core/test/bundling/todo/index.html]: done -> gone

**Review by kara — APPROVED** [review]
LGTM

File: tools/public_api_guard/core/core.d.ts [hunk 2]
```diff
@@ -370,7 +370,8 @@ export declare const HostListener: HostListenerDecorator;
 
 /** @experimental */
 export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue?: undefined, flags?: InjectFlags): T;
-export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue: T | null, flags?: InjectFlags): T | null;
+export declare function inject<T>(token: Type<T> | InjectionToken<T>, notFoundValue: T, flags?: InjectFlags): T;
```
Comment (mhevery) [inline tools/public_api_guard/core/core.d.ts]: This change is needed to otherwise `inject(Foo, new Foo())`'s type is `Foo|null` which is not correct.

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 3]
```diff
@@ -762,7 +751,7 @@ export function createFactory(
       unsupported('value dependencies');
     }
     if (dependency.isHost) {
-      unsupported('host dependencies');
+      // unsupported('host dependencies');
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: why ? add a comment ?

Comment (mhevery) [reply packages/compiler/src/render3/r3_view_compiler.ts]: Opps, will revert

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 4]
```diff
@@ -599,17 +599,6 @@ class TemplateDefinitionBuilder implements TemplateAstVisitor, LocalResolver {
             this._bindingMode, directive.sourceSpan, R3.elementProperty, o.literal(nodeIndex),
             o.literal(input.templateName), o.importExpr(R3.bind).callFn([convertedBinding]));
       }
-
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: Should those changes be part of a different commit describing the rationale for the change ?

File: packages/core/test/bundling/todo/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+ng_rollup_bundle(
+    name = "bundle",
+    # TODO(alexeagle): This is inconsistent.
+    # We try to teach users to always have their workspace at the start of a
+    # path, to disambiguate from other workspaces.
+    # Here, the rule implementation is looking in an execroot where the layout
+    # has an "external" directory for external dependencies.
+    # This should probably start with "angular/" and let the rule deal with it.
```
Comment (vicb) [inline packages/core/test/bundling/todo/BUILD.bazel]: `# TODO: Remove once #22913 lands` ?

File: packages/core/test/bundling/todo/index.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  done: boolean;
+}
+
+@Injectable({providedIn: 'root'})
+export class AppState {
+  todos: ToDo[] = [
+    {text: 'Demonstrate Components', done: false},
+    {text: 'Demonstrate Structural Directives', done: true},
+    {text: 'Demonstrate NgModules', done: false},
+    {text: 'Demonstrate zoneless changed detection', done: false},
```
Comment (vicb) [inline packages/core/test/bundling/todo/index.ts]: typo "change"

File: packages/core/test/bundling/todo/index.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  static DEFAULT_TODO = {text: '', done: false};
+}
+
+@Component({
+  selector: 'todo',
+  // TODO(misko): `[class.done]` and `[value]` should be `todo.done` not `todo && todo.todo.done`
+  // The reason for the guard is that template executes creation and binding together
+  // but NgForOf expects creation and binding separate.
+  template: `
+    <div [class.done]="todo && todo.done">
```
Comment (vicb) [inline packages/core/test/bundling/todo/index.ts]: `todo?.done` ?

File: packages/core/test/bundling/todo/index.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  </div>
+  <span>count: {{appState.todos.length}}.</span>
+  `
+})
+export class ToDoAppComponent {
+  public appState: AppState;
+
+  // TODO(misko): Injection is broken because the compiler generates incorrect code.
+  constructor(/**appState: AppState*/) {
+    // TODO(misko): remove once injection is working.
```
Comment (vicb) [inline packages/core/test/bundling/todo/index.ts]: duplicate TODO ?

**Review by vicb — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]
lgtm for api changes

but I'm surprised to see that we now have ivy_ng_module - fragmenting the build system like this seem like a bad idea. did you consult this with @alexeagle ?

**Comment by mhevery** [comment]
@IgorMinar Yes it is a bad idea. 🙂  It is internal only build rule/hack until we have proper rule https://github.com/angular/angular/blob/master/packages/compiler/design/separate_compilation.md#ng_module-output-bazel and yes this was discussed with @alexeagle @alxhub and @chuckjaz

File: packages/core/test/bundling/todo/index.html (hunk 1 above)
Comment (mhevery) [reply packages/core/test/bundling/todo/index.html]: fixed

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 4 above)
Comment (mhevery) [reply packages/compiler/src/render3/r3_view_compiler.ts]: broken to separate commits

File: packages/core/test/bundling/todo/BUILD.bazel (hunk 5 above)
Comment (mhevery) [reply packages/core/test/bundling/todo/BUILD.bazel]: https://github.com/angular/angular/pull/22913 is unrelated to this.

File: packages/core/test/bundling/todo/index.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/test/bundling/todo/index.ts]: fixed

File: packages/core/test/bundling/todo/index.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/test/bundling/todo/index.ts]: I intentionally left it in this ugly state because it is a hack which should be removed once we fix it in ivy runtime.

File: packages/core/test/bundling/todo/index.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  template: `
+    <div [class.done]="todo && todo.done">
+      <input type="checkbox" [value]="todo && todo.done" (click)="onCheckboxClick()">
+      <span>{{todo && todo.text}}</span>
+      <button (click)="onArchiveClick()">archive</button>
+    </div>
+  `
+})
+export class TodoComponent {
+  static DEFAULT_TODO: ToDo = {text: '', done: false};
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/bundling/todo/index.ts]: removed

File: packages/core/test/bundling/todo/index.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/test/bundling/todo/index.ts]: fixed

File: packages/core/test/bundling/todo/BUILD.bazel (hunk 5 above)
Comment (vicb) [reply packages/core/test/bundling/todo/BUILD.bazel]: Well... that comes from you: https://github.com/angular/angular/pull/22654#discussion_r176233174

What did you mean then ?

**Review by kara — APPROVED** [review]
LGTM

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts [hunk 10]
```diff
@@ -211,23 +207,19 @@ describe('compiler compliance', () => {
...
         static ngComponentDef = $r3$.ɵdefineComponent({
           type: MyComponent,
           selector: [[['my-component'], null]],
           factory: function MyComponent_Factory() { return new MyComponent(); },
           template: function MyComponent_Template(ctx: IDENT, cm: IDENT) {
             if (cm) {
-              $r3$.ɵE(0, 'ul', null, null, $c1$);
-              $r3$.ɵC(2, $c2$, MyComponent_IfDirective_Template_2);
+              $r3$.ɵE(0, 'ul', null, $c1$);
+              $r3$.ɵC(2, MyComponent_IfDirective_Template_2, null, $c2$);
```
Comment (trotyl) [inline packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: Since `IfDirective` is not determined at this stage any more, how could it still remains in the template function name? Same for `MyComponent_ForOfDirective_Template_1` below.

File: packages/core/src/render3/instructions.ts [hunk 11]
```diff
@@ -400,7 +400,8 @@ function resetApplicationState() {
  * @param context to pass into the template.
  * @param providedRendererFactory renderer factory to use
  * @param host The host element node to use
- * @param defs Any directive or pipe defs that should be used for matching
+ * @param directives Any directive defs that should be used for matching
+ * @param pipes Any pipe defs that should be used for matching
```
Comment (trotyl) [inline packages/core/src/render3/instructions.ts]: `Any directive defs` -> `Any directive types` according to the change? (Should `any` be combined with plural?)

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts (hunk 10 above)
Comment (mhevery) [reply packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: You are right it should not be in the function name. We currently are still using old backend for the compiler which still has this information. The new compiler pipeline `ngtsc` will not and so when we switch over we will not be able to generate the string. Since the function name is there just to aid in debugging, i am going to leave it for now since it will get fixe automatically when we switch over to `ngtsc`.

File: packages/core/src/render3/instructions.ts (hunk 11 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: Removed the word `any`. Yes it is deys not types. We pass types to `defineComponent` which in turn converts them to defs. Because going from Type to Defs is megamorphic we want to do it as early as possible and than have all of the runtime be defs only.
